File: 0121.BestTimetoBuyandSellStock/BestTimetoBuyandSellStock.go
/*
Time complexity: O(N)
Space complexity: O(1)

Runtime Faster than 99%+
*/

func maxProfit(prices []int) int {
    result := 0
    lastData := prices[0]
    tmp := 0
    for i := 0; len(prices) > i; i++ {
        if lastData > prices[i]{
            lastData = prices[i]
        } else {
            tmp = prices[i] - lastData
            if result < tmp {
                result = tmp
            }
        }
    }
    return result
}

File: 0121.BestTimetoBuyandSellStock/BestTimetoBuyandSellStock.py
'''
Time complexity: O(N)
Space complexity: O(1)

Runtime Faster than 90%+
'''


class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        min_p, result = prices[0], 0
        for price in prices:
            if price <= min_p:
                min_p = price
            else:
                result = max(price-min_p, result)
        return result
